File: bench/binarytrees/binarytrees.pl
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org/
#
# contributed by Emanuele Zeppieri
# modified by Christian Walde (threads)

use threads;

sub bottomup_tree {
    my ( $value, $depth ) = @_;
    return $value unless $depth;
    my $value2 = $value * 2;
    $depth--;
    return [ bottomup_tree( $value2 - 1, $depth ), bottomup_tree( $value2, $depth ), $value ];
}

sub check_tree {
    my ( $left, $right, $value ) = @{ $_[0] };
    return $value + ( ref $left ? check_tree( $left ) - check_tree( $right ) : $left - $right );
}

sub depth_iteration {
    my ( $depth, $max_depth, $min_depth ) = @_;

    my $iterations = 2 << $max_depth - $depth + $min_depth - 1;
    my $check      = 0;

    foreach ( 1 .. $iterations ) {
        $check += check_tree( bottomup_tree( $_,  $depth ) );
        $check += check_tree( bottomup_tree( -$_, $depth ) );
    }

    return ( $depth => [ $iterations, $depth, $check ] );
}

my $max_depth = shift @ARGV;
my $min_depth = 4;

$max_depth = $min_depth + 2 if $min_depth + 2 > $max_depth;

my $stretch_depth = $max_depth + 1;
my $stretch_tree = bottomup_tree( 0, $stretch_depth );
print "stretch tree of depth $stretch_depth\t check: ", check_tree( $stretch_tree ), "\n";
undef $stretch_tree;

my $longlived_tree = bottomup_tree( 0, $max_depth );

my @results;
for ( my $depth = $min_depth ; $depth <= $max_depth ; $depth += 2 ) {
    while ( 1 ) {
        last if threads->list < 4;
        push @results, $_->join for threads->list( threads::joinable );
    }

    threads->create(
        { 'context' => 'list', 'stack_size' => 64 },    #
        sub { depth_iteration( $depth, $max_depth, $min_depth ) }
    );
}

while ( threads->list ) {
    push @results, $_->join for threads->list( threads::joinable );
}

my %results = @results;
for my $key ( sort { $a <=> $b } keys %results ) {
    my ( $iterations, $depth, $check ) = @{ $results{$key} };
    print 2 * $iterations, "\t trees of depth $depth\t check: ", $check, "\n";
}

print "long lived tree of depth $max_depth\t check: ", check_tree( $longlived_tree ), "\n";

File: bench/binarytrees/binarytrees.go
/* The Computer Language Benchmarks Game
 * http://benchmarksgame.alioth.debian.org/
 *
 * contributed by Ugorji Nwoke
 * *reset*
 */

package main

import (
	"fmt"
	"os"
	"strconv"
)

const minDepth = 4

type treeNode struct {
	left, right *treeNode
}

func (n *treeNode) itemCheck() int {
	if n.left == nil {
		return 1
	}
	return 1 + n.left.itemCheck() + n.right.itemCheck()
}

func bottomUp(depth int) *treeNode {
	if depth > 0 {
		return &treeNode{
			bottomUp(depth - 1),
			bottomUp(depth - 1),
		}
	}
	return &treeNode{nil, nil}
}

func main() {
	n := 0
	if len(os.Args) > 1 {
		if n2, err2 := strconv.ParseInt(os.Args[1], 10, 0); err2 == nil {
			n = int(n2)
		}
	}
	maxDepth := n
	if minDepth+2 > n {
		maxDepth = minDepth + 2
	}
	stretchDepth := maxDepth + 1
	check := bottomUp(stretchDepth).itemCheck()
	fmt.Printf("stretch tree of depth %v\t check: %v\n", stretchDepth, check)
	longLivedTree := bottomUp(maxDepth)
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		interactions := 1 << uint(maxDepth-depth+minDepth)
		check = 0
		for i := 1; i <= interactions; i++ {
			check += bottomUp(depth).itemCheck()
		}
		fmt.Printf("%v\t trees of depth %v\t check: %v\n", interactions, depth, check)
	}
	fmt.Printf("long lived tree of depth %v\t check: %v\n", maxDepth, longLivedTree.itemCheck())
}
